python_modules/dagster/dagster/_core/definitions/asset_graph_subset.py

```diff
@@ -1,70 +1,129 @@
+        ],
+    )
+):
+    def __new__(
+        cls,
+        partitions_subsets_by_serialized_asset_key: Optional[Mapping[str, PartitionsSubset]] = None,
         non_partitioned_asset_keys: Optional[AbstractSet[AssetKey]] = None,
+        partitions_subsets_by_asset_key: Optional[Mapping[AssetKey, PartitionsSubset]] = None,
     ):
-        self._asset_graph = asset_graph
-        self._partitions_subsets_by_asset_key = partitions_subsets_by_asset_key or {}
-        self._non_partitioned_asset_keys = non_partitioned_asset_keys or set()
+        check.invariant(
+            not (partitions_subsets_by_serialized_asset_key and partitions_subsets_by_asset_key),
+            "Cannot provide both partitions_subsets_by_serialized_asset_key and partitions_subsets_by_asset_key",
+        )
 
-    @property
-    def asset_graph(self) -> AssetGraph:
-        return self._asset_graph
+        if partitions_subsets_by_asset_key:
+            partitions_subsets_by_serialized_asset_key = {
+                key.to_user_string(): value
+                for key, value in partitions_subsets_by_asset_key.items()
+            }
+
+        return super(AssetGraphSubset, cls).__new__(
+            cls,
+            partitions_subsets_by_serialized_asset_key=partitions_subsets_by_serialized_asset_key
+            or {},
+            non_partitioned_asset_keys=non_partitioned_asset_keys or set(),
+        )
 
-    @property
+    @cached_property
     def partitions_subsets_by_asset_key(self) -> Mapping[AssetKey, PartitionsSubset]:
-        return self._partitions_subsets_by_asset_key
-
-    @property
-    def non_partitioned_asset_keys(self) -> AbstractSet[AssetKey]:
-        return self._non_partitioned_asset_keys
+        return {
+            AssetKey.from_user_string(key): value
+            for key, value in self.partitions_subsets_by_serialized_asset_key.items()
+        }
 
     @property
     def asset_keys(self) -> AbstractSet[AssetKey]:
         return {
             key for key, subset in self.partitions_subsets_by_asset_key.items() if len(subset) > 0
-        } | self._non_partitioned_asset_keys
+        } | self.non_partitioned_asset_keys
 
     @property
     def num_partitions_and_non_partitioned_assets(self):
-        return len(self._non_partitioned_asset_keys) + sum(
-            len(subset) for subset in self._partitions_subsets_by_asset_key.values()
+        return len(self.non_partitioned_asset_keys) + sum(
+            len(subset) for subset in self.partitions_subsets_by_asset_key.values()
         )
 
-    def get_partitions_subset(self, asset_key: AssetKey) -> PartitionsSubset:
-        partitions_def = self.asset_graph.get_partitions_def(asset_key)
-        if partitions_def is None:
-            check.failed("Can only call get_partitions_subset on a partitioned asset")
+    def get_partitions_subset(
+        self, asset_key: AssetKey, asset_graph: Optional[AssetGraph] = None
```

Comment:
passed in asset graph because we'd like to get an empty subset instead of None if possible

the `__oper__` callsite does not have access to the asset graph, so this is an optional param